src/group_theory/nilpotent.lean

```diff
@@ -871,3 +854,48 @@ begin
     apply map_injective_of_ker_le (mk' (center G)) hkh le_top,
     exact (ih H' hH').trans (symm (map_top_of_surjective _ hsur)), },
 end
+
+end with_group
+
+section with_finite_group
+
+open group
+
+variables {G : Type*} [hG : group G] [hf : fintype G]
+include hG hf
+
+/-- A p-group is nilpotent -/
+lemma is_p_group.is_nilpotent {p : ℕ} [hp : fact (nat.prime p)] (h : is_p_group p G) :
+  is_nilpotent G :=
+begin
+  classical,
+  unfreezingI
+  { revert hG,
+    induction hf using fintype.induction_subsingleton_or_nontrivial with G hG hS G hG hN ih },
+  { apply_instance, },
+  { introI _, intro h,
+    have hc : center G > ⊥ := gt_iff_lt.mp h.bot_lt_center,
+    have hcq : fintype.card (G ⧸ center G) < fintype.card G,
+    { rw card_eq_card_quotient_mul_card_subgroup (center G),
+      apply lt_mul_of_one_lt_right,
+      exact (fintype.card_pos_iff.mpr has_one.nonempty),
+      exact ((subgroup.one_lt_card_iff_ne_bot _).mpr (ne_of_gt hc)), },
+    have hnq : is_nilpotent (G ⧸ center G) := ih _ hcq (h.to_quotient (center G)),
+    exact (of_quotient_center_nilpotent hnq), }
+end
+
+/-- If a finite group is the direct product of its Sylow groups, it is nilpotent -/
+theorem is_nilpotent_of_product_of_sylow_group
```

Comment:
Would it be easy to generalize this to: A finite product of finite p-groups is nilpotent?
Then this would just be a special case.

Reply:
I guess I do want this one as a statement for https://github.com/leanprover-community/mathlib/pull/11835. I can put the more general one as a lemma above, but it'd be itself just a simple composition of two lemmas.

Reply:
But of course no problem factoring this into two lemmas. Will do!

Reply:
You mean consistent with https://github.com/leanprover-community/mathlib/pull/11778/files#diff-ab70e1f803f9921c903cc935dd827a15a25f8323fad19b349c8f756bb396c059R593? Yes, of course, else I couldn’t proof all these characterizations equivalent :-)